src/preferences.c

```diff
@@ -15,193 +15,50 @@
-  if(!setting || config_setting_type(setting) != type){
-    config_setting_remove(root, key);
-    setting = config_setting_add(root, key, type);
-    config_setting_set_bool(setting, default_value);
-  }
-  return setting;
-}
-
-config_setting_t* preferences_init_group(config_setting_t* root, char* key){
-  int type = CONFIG_TYPE_GROUP;
-  config_setting_t* setting = config_setting_get_member(root, key);
-  if(!setting || config_setting_type(setting) != type){
-    config_setting_remove(root, key);
-    setting = config_setting_add(root, key, type);
-  }
-  return setting;
-}
-
-char preferences_is_list_of_groups_of_strings(config_setting_t* setting){
-  config_setting_t* subsetting;
-
-	if(config_setting_is_list(setting) != CONFIG_TRUE){
-		return 0;
-	}
-  int num_elements = config_setting_length(setting); // num elements in list
-  int num_subelements; // num elements in each group in the list
-  unsigned int i, j;
-  for(i = 0; i < num_elements; ++i){
-  	subsetting = config_setting_get_elem(setting, i);
-  	if(config_setting_is_group(subsetting) != CONFIG_TRUE){
-  		return 0;
-  	}
-  	// now check for all string values
-  	num_subelements = config_setting_length(subsetting);
-  	for(j = 0; j < num_subelements; ++j){
-  		if(config_setting_type(config_setting_get_elem(subsetting, j)) != CONFIG_TYPE_STRING){
-  			return 0;
-  		}
-  	}
-  }
-  return 1;
-}
-
-config_setting_t* preferences_init_list_blank(config_setting_t* root, char* key){
-  int type = CONFIG_TYPE_LIST;
-  config_setting_t* setting = config_setting_get_member(root, key);
-  if(!setting || config_setting_type(setting) != type || config_setting_length(setting) > 0){
-    config_setting_remove(root, key);
-    setting = config_setting_add(root, key, type);
-  }
-  return setting;
-}
-
-config_setting_t* preferences_init_array(config_setting_t* root, char* key){
-  int type = CONFIG_TYPE_ARRAY;
-  config_setting_t* setting = config_setting_get_member(root, key);
-  if(!setting || config_setting_type(setting) != type){
-    config_setting_remove(root, key);
-    setting = config_setting_add(root, key, type);
-  }
-  return setting;
-}
-
-void preferences_pad_array_int(config_setting_t* setting, int size, int padding){
-	/* we assume that setting is actually a CONFIG_TYPE_ARRAY */
-	int num = config_setting_length(setting);
-	for(; num < size; ++num){
-		config_setting_set_int_elem(setting, -1, padding);
+pref_t *prefs = NULL;
+
+static void first_run(pref_t *prefs) {
+	char* home = getenv("HOME");
+	if(home != NULL){ chdir(home); }
+	
+	char* readme_path = (atoi(getenv("WIDTH")) <= 720) ? README45_FILE_PATH : README_FILE_PATH;
+	fprintf(stderr, "Updating README\n");
+	if (access(readme_path, F_OK) != -1) {
+		// stat success!
+		if (symlink(readme_path, "./README") == -1){
+			if (errno != EEXIST){
+				fprintf(stderr, "Error linking README from app to PWD\n");
+			}
+		}
 	}
-}
 
-int preferences_check_version(config_setting_t* root, char* key){
```

Comment:
This check needs to stay. Without it, people upgrading from earlier versions will have their old style preferences silently ignored. Currently, when I take a V8 config file and copy it into this new version, my custom preferences for metamode keys and such are silently ignored, because the type for metamode_keys has changed from group to list. On launch, we need to detect old preference versions and convert them for the new version, then update the preferences version and write a new file.